File: python/mirascope/llm/providers/together/provider.py
"""Together AI provider implementation."""

from typing import ClassVar

from ..openai.completions.base_provider import BaseOpenAICompletionsProvider


class TogetherProvider(BaseOpenAICompletionsProvider):
    """Provider for Together AI's OpenAI-compatible API.

    Inherits from BaseOpenAICompletionsProvider with Together-specific configuration:
    - Uses Together AI's API endpoint
    - Requires TOGETHER_API_KEY

    Usage:
        Register the provider with model ID prefixes you want to use:

        ```python
        import llm

        # Register for meta-llama models
        llm.register_provider("together", "meta-llama/")

        # Now you can use meta-llama models directly
        @llm.call("meta-llama/Llama-3.3-70B-Instruct-Turbo")
        def my_prompt():
            return [llm.messages.user("Hello!")]
        ```
    """

    id: ClassVar[str] = "together"
    default_scope: ClassVar[str | list[str]] = []
    default_base_url: ClassVar[str | None] = "https://api.together.xyz/v1"
    api_key_env_var: ClassVar[str] = "TOGETHER_API_KEY"
    api_key_required: ClassVar[bool] = True
    provider_name: ClassVar[str | None] = "Together"

    def _model_name(self, model_id: str) -> str:
        """Return the model ID as-is for Together API."""
        return model_id

File: typescript/src/llm/providers/together/provider.ts
/**
 * Together AI provider implementation.
 *
 * Together AI provides an OpenAI-compatible API for hosted LLM inference.
 */

import { BaseOpenAICompletionsProvider } from '@/llm/providers/openai/completions/base-provider';

const DEFAULT_BASE_URL = 'https://api.together.xyz/v1';
const API_KEY_ENV_VAR = 'TOGETHER_API_KEY';

export interface TogetherProviderInit {
  apiKey?: string;
  baseURL?: string;
}

/**
 * Provider for Together AI's OpenAI-compatible API.
 *
 * Inherits from BaseOpenAICompletionsProvider with Together-specific configuration:
 * - Uses Together AI's API endpoint (https://api.together.xyz/v1)
 * - Requires TOGETHER_API_KEY environment variable
 *
 * @example
 * ```typescript
 * import { registerProvider } from '@mirascope/core/llm';
 *
 * // Register for meta-llama models
 * registerProvider('together', 'meta-llama/');
 *
 * // Now you can use meta-llama models directly
 * const call = defineCall({
 *   model: 'meta-llama/Llama-3.3-70B-Instruct-Turbo',
 *   fn: () => [user('Hello!')],
 * });
 * ```
 */
export class TogetherProvider extends BaseOpenAICompletionsProvider {
  readonly id = 'together' as const;

  /**
   * Create a new Together AI provider instance.
   *
   * @param init - Configuration options
   * @param init.apiKey - API key (defaults to TOGETHER_API_KEY env var)
   * @param init.baseURL - Base URL (defaults to 'https://api.together.xyz/v1')
   */
  constructor(init: TogetherProviderInit = {}) {
    const resolvedApiKey = init.apiKey ?? process.env[API_KEY_ENV_VAR];
    const resolvedBaseUrl = init.baseURL ?? DEFAULT_BASE_URL;

    super({
      id: 'together',
      apiKey: resolvedApiKey,
      baseURL: resolvedBaseUrl,
    });
  }
}
